Karma: Don't create duplicate entries for names that differ only in lower- and uppercase

salto-edit opened the pull request:
Suppose someone upvoted "LtWorf" for the first time like so

```
<someone> LtWorf++
<LtData> Prendo nota. LtWorf(1)
```

And later someone upvotes "ltworf" because she is too lazy to spell LtWorf's nickname right, then

```
<someone> ltworf++
<LtData> Prendo nota. LtWorf(2)
```

and the updated database will not contain the key "ltworf" but "LtWorf" with an incremented karma counter.

salto-edit committed:
Fix typo

salto-edit committed:
Factor out up- and downvote code to vote function

salto-edit committed:
Amend Factor out up- and downvote code to vote function

- move database update to the end of `vote`
- output new karma when bot is upvoted.

salto-edit committed:
Normalise keys when searching karma database

- the name of a new entry is still taken literally

salto-edit commented:
Commits bea77ab and 2b79704 can be squashed.

ltworf commented on modules/karma.py:
```diff
@@ -70,28 +70,29 @@ def sendmsg(sender, recip, text):
+        return vote(text[:-2], -1)
     return None
 
+def vote(nick, delta=1):
```
I don't think that a optional parameter is good here.

salto-edit replied:
You find the interface confusing? It's not needed, so go ahead and make it a parameter.

ltworf commented on modules/karma.py:
```diff
@@ -70,28 +70,29 @@ def sendmsg(sender, recip, text):
+        result = "Grazie per la tua stima"
+    elif nick.lower() == config['nickname'].lower() and delta <= 0:
+        return "Nah, non credo di volerlo fare"
+    elif nick.lower() == sender.lower() and delta > 0:
```
sender is not defined

salto-edit replied:
See follow up commit 1aae5ff.

salto-edit committed:
Fixup Factor out up- and downvote code to vote function

Instead of adding `sender` to the context of `vote`, break the vote control flow in `sendmsg` already when trying to upvote oneself or downvote us.

salto-edit committed:
Fixup Normalise keys when searching karma database

- store representation of the normalised key in the database
- store database as JSON on disk to avoid cumbersome and ad-hoc save/load format

salto-edit commented:
Here you go, don't forget to convert the database again. :tada:

ltworf commented on modules/karma.py:
```diff
@@ -28,28 +29,22 @@ def init():
-    for i in karma:
-        f.write("%s#%d" % (i, karma[i]))
-        f.write("\n")
+    json.dump(f)
```
you forgot the data that you want to dump here

salto-edit replied:
Sepyt!
